types/module/server/module_key.go

```diff
@@ -1,20 +1,99 @@
 package server
 
 import (
+	"context"
+	"fmt"
+
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	"google.golang.org/grpc"
+
 	"github.com/regen-network/regen-ledger/types"
+	"github.com/regen-network/regen-ledger/types/address"
 )
 
+// ModuleKey is an interface for module servers required by router.
 type ModuleKey interface {
 	types.InvokerConn
 
 	ModuleID() types.ModuleID
 	Address() sdk.AccAddress
+	Derive(key []byte) ModuleKey
+}
+
+type RootModuleKey interface {
+	ModuleKey
+	sdk.StoreKey
+}
+
+type moduleKey struct {
+	moduleName string
+	addr       []byte
+	i          InvokerFactory
+}
+
+// NewDerivedModuleKey creates a ModuleKey with a derived moduel address based on parent
+// module address and derivation key.
+func NewDerivedModuleKey(modName string, parentAddr, derivationKey []byte, i InvokerFactory) ModuleKey {
```

Comment:
why is this public?

Reply:
Because the original structure - `DerivedModuleKey` - was public, so I though you want to have it exposed.